### ISSUE TITLE
function definition should allow compound command

### ISSUE THREAD

**Issue by mmjo** [koalaman/shellcheck#2130]
#### For bugs
- Rule Id: SC1073, SC1064, SC1072
- My shellcheck version (`shellcheck --version` or 'online'): online
- [x] I tried on shellcheck.net and verified that this is still a problem on the latest commit

#### For new checks and feature suggestions
- [x] shellcheck.net (i.e. the latest commit) currently gives no useful warnings about this
- [x] I searched through https://github.com/koalaman/shellcheck/issues and didn't find anything related

#### Here's a snippet or screenshot that shows the problem:

```sh
#!/bin/bash
f() [[ 1 = 1 ]]
```

#### Here's what shellcheck currently says:
```
Line 2	SC1073: Couldn't parse this function. Fix to allow more checks.
Line 2	SC1064: Expected a { to open the function definition.
Line 2	SC1072:  Fix any mentioned problems and try again.
```
#### Here's what I wanted or expected to see:
No warnings.

#### Musings
It seems from `src/Parser.hs` in `readFunctionDefinition`
that shellcheck only allows f() {...} and f() (...).

Though both bash and posix allow a compound command as function body.

Note that the `readCompoundCommand` could have been useful.
But unfortunately it includes `readFunctionDefinition` itself.

**Comment by matthewpersico** [comment]
What would this do? If you try something useful:
```
 $ f() echo hello
-bash: syntax error near unexpected token `echo'
```

**Comment by mmjo** [comment]
'echo hello' is not a compound command, so the error is expected.

A slightly more useful example:
`is_ok () [[ "$1" = 42 ]]`

Compare against the verbose version (no need for a return, but do not forget the semicolon):
`is_ok () { [[ "$1" = 42 ]]; }`

You can even write:
`is_ok () if [ "$1" = 42 ]; then return 0; else return 1; fi`

Instead of an 'if'-statement, a case/while/until/select/for statement is also possible.
[but these i never use]

Last possibility is arithmetic. E.g.
`is_ok () (( "$1" - 42 == 0 ))`

**Comment by matthewpersico** [comment]
TIL

**Comment by TinCanTech** [comment]
> Last possibility is arithmetic. E.g.
> `is_ok () (( "$1" - 42 == 0 ))`

According to Douglas Adams, _The Answer_ is not 42 because _The Question_ is:<br>
- _What is five times nine_ ?

42 was a Deep-Thought error. Which was the reason for Earth.

**Comment by koalaman** [comment]
> Note that the `readCompoundCommand` could have been useful.
> But unfortunately it includes `readFunctionDefinition` itself.

This is not a problem for `dash`:

```
$ f() f() { echo foo; }
$ f
$ f
foo
```

The original reason for not allowing anything other than {} and () was that ShellCheck kept giving misleading advice when it interpreted bad array assignments and similar as possible function definitions. However, a lot of that parsing was rewritten in 0.7.2, so it may be feasible to allow this again.

**Comment by dermoth** [comment]
> Note that the `readCompoundCommand` could have been useful. But unfortunately it includes `readFunctionDefinition` itself.

And it has to, a function can be defined within a function. Can't this be recursive?
